documented-macros/src/lib.rs

```diff
@@ -407,38 +410,95 @@ pub fn documented_variants(input: TokenStream) -> TokenStream {
-        .filter_map(|attr| match attr.meta {
-            Meta::NameValue(ref name_value) if name_value.path.is_ident("doc") => {
-                Some(&name_value.value)
-            }
-            _ => None,
-        })
-        .map(|expr| match expr {
-            Expr::Lit(ExprLit { lit: Lit::Str(s), .. }) => Ok(s.value()),
-            other => Err(Error::new(
-                other.span(),
-                "Doc comment is not a string literal",
-            )),
-        })
-        .collect::<Result<Vec<_>, _>>()?;
+/// Macro to extract the documentation on any item that accepts doc comments
+/// and store it in a const variable.
+///
+/// # Examples
+///
+/// ```rust
+/// use documented::docs_const;
+///
+/// /// This is a test function
+/// #[docs_const]
+/// fn test_fn() {}
+///
+/// assert_eq!(TEST_FN_DOCS, "This is a test function");
+/// ```
+///
+/// # Configuration
+///
+/// With the `customise` feature enabled, you can customise this macro's
+/// behaviour using attribute arguments.
+///
+/// Currently, you can:
+///
+/// ## 1. set a custom constant name like so:
+///
+/// ```rust
+/// use documented::docs_const;
+///
+/// /// If you have a question raise your hand
+/// #[docs_const(name = "DONT_RAISE_YOUR_HAND")]
+/// mod whatever {}
+///
+/// assert_eq!(DONT_RAISE_YOUR_HAND, "If you have a question raise your hand");
+/// ```
+///
+/// ## 2. disable line-trimming like so:
+///
+/// ```rust
+/// use documented::docs_const;
+///
+/// ///     This is a test constant
+/// #[docs_const(trim = false)]
+/// const test_const: u8 = 0;
+///
+/// assert_eq!(TEST_CONST_DOCS, "     This is a test constant");
+/// ```
+///
+/// ---
+///
+/// Multiple option can be specified in a list like so:
+/// `name = "FOO", trim = false`.
+///
+/// If there are other configuration options you wish to have, please
+/// submit an issue or a PR.
+#[proc_macro_attribute]
+pub fn docs_const(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let item = syn::parse_macro_input!(item as Item);
 
-    if string_literals.is_empty() {
-        return Ok(None);
-    }
+    #[cfg(not(feature = "customise"))]
+    let config = AttrConfig::default();
+    #[cfg(feature = "customise")]
+    let config = AttrConfig::default()
+        .with_customisations(syn::parse_macro_input!(attr as AttrCustomisations));
 
-    let docs = if config.trim {
-        string_literals
-            .iter()
-            .flat_map(|lit| lit.split('\n').collect::<Vec<_>>())
-            .map(|line| line.trim().to_string())
-            .collect::<Vec<_>>()
-            .join("\n")
-    } else {
-        string_literals.join("\n")
+    let (item_vis, item_name, attrs) = match get_vis_name_attrs(&item) {
+        Ok(pair) => pair,
+        Err(e) => return e.into_compile_error().into(),
     };
 
-    Ok(Some(docs))
+    let docs = match get_docs(attrs, config.trim) {
+        Ok(Some(docs)) => docs,
+        Ok(None) => {
+            // IDEA: customisation: allow_empty
```

Comment:
This option would be applicable to the derive macros as well, so I'll leave this here for now.